kdump.conf.5
.TH KDUMP.CONF 5 "07/23/2008" "kdump-utils"

.SH NAME
kdump.conf \- configuration file for kdump kernel.

.SH DESCRIPTION

kdump.conf is a configuration file for the kdump kernel crash
collection service.

kdump.conf provides post-kexec instructions to the kdump kernel. It is
stored in the initrd file managed by the kdump service. If you change
this file and do not want to reboot in order for the changes to take
effect, restart the kdump service to rebuild the initrd.

For most configurations, you can simply review the examples provided
in the stock /etc/kdump.conf.

.B NOTE:
For filesystem dumps the dump target must be mounted before building
kdump initramfs.

kdump.conf only affects the behavior of the initramfs.  Please read the
kdump operational flow section of kexec-kdump-howto.txt in the docs to better
understand how this configuration file affects the behavior of kdump.

.SH OPTIONS

.B auto_reset_crashkernel <yes|no>
.RS
determine whether to reset kernel crashkernel parameter to the default value
or not when kdump-utils is updated or a new kernel is installed. The default
crashkernel values are different for different architectures and also take the
following factors into consideration,
.IP
\- AMD Secure Memory Encryption (SME) and Secure Encrypted Virtualization (SEV)
.IP
\- Mellanox 5th generation network driver
.IP
\- aarch64 64k kernel
.IP
\- Firmware-assisted dump (FADump)
.PP
Since the kernel crasherkernel parameter will be only reset when kdump-utils is
updated or a new kernel is installed, you need to call "kdumpctl
reset-crashkernel [--kernel=path_to_kernel]" if you want to use the default
value set after having enabled features like SME/SEV for a specific kernel. And
you should also reboot the system for the new crashkernel value to take effect.
Also see kdumpctl(8).
.PP
.RE

.B raw <partition>
.RS
Will dd /proc/vmcore into <partition>.  Use persistent device names for
partition devices, such as /dev/vg/<devname>. The "raw" dump target is
deprecated and will be removed in a future release. Please migrate to a
filesystem-based dump target (e.g., ext4, xfs, nfs, or ssh).
.RE

.B nfs <nfs mount>
.RS
Will mount nfs to <mnt>, and copy /proc/vmcore to <mnt>/<path>/%HOST-%DATE/,
supports DNS. Note that a fqdn should be used as the server name in the
mount point.
.RE

.B ssh <user@server>
.RS
Will save /proc/vmcore through ssh pipe to <user@server>:<path>/%HOST-%DATE/,
supports DNS. NOTE: make sure user has necessary write permissions on
server and that a fqdn is used as the server name.
.RE

.B sshkey <path>
.RS
Specify the path of the ssh key to use when dumping via ssh.
The default value is /root/.ssh/kdump_id_rsa.
.RE

.B <fs type> <partition>
.RS
Will mount -t <fs type> <partition> <mnt>, and copy /proc/vmcore to
<mnt>/<path>/%HOST_IP-%DATE/.  NOTE: <partition> can be a device node, label
or uuid.  It's recommended to use persistent device names such as
/dev/vg/<devname>.  Otherwise it's suggested to use label or uuid.
.RE

.B path <path>
.RS
"path" represents the file system path in which vmcore will be saved.
If a dump target is specified in kdump.conf, then "path" is relative to the
specified dump target.
.PP
Interpretation of "path" changes a bit if the user didn't specify any dump
target explicitly in kdump.conf. In this case, "path" represents the
absolute path from root.  The dump target and adjusted path are arrived
at automatically depending on what's mounted in the current system.
.PP
Ignored for raw device dumps.  If unset, will use the default "/var/crash".
.RE

.B core_collector <command> <options>
.RS
This allows you to specify the command to copy the vmcore.
The default is makedumpfile, which on some architectures can drastically reduce
core file size. See /sbin/makedumpfile --help for a list of options.
Note that the -i and -g options are not needed here, as the initrd
will automatically be populated with a config file appropriate
for the running kernel.
.PP
Note 1: About default core collector:
The default core_collector for raw/ssh dump is:
"makedumpfile -F -l --message-level 7 -d 31".
The default core_collector for other targets is:
"makedumpfile -l --message-level 7 -d 31".
Even if core_collector option is commented out in kdump.conf, makedumpfile
is the default core collector and kdump uses it internally.
If one does not want makedumpfile as default core_collector, then they
need to specify one using core_collector option to change the behavior.
.PP
Note 2: If "makedumpfile -F" is used then you will get a flattened format
vmcore.flat, you will need to use "makedumpfile -R" to rearrange the
dump data from standard input to a normal dumpfile (readable with analysis
tools).
ie. "makedumpfile -R vmcore < vmcore.flat"
.PP
Note 3: If specified core_collector simply copy the vmcore file to the
dump target (eg: cp, scp), the vmcore could be significantly large.
Please make sure the dump target has enough space, at leaset larger
than the system's RAM.

.RE

.B kdump_emergency <binary | script>
.RS
This directive allows you to run a specified executable
in emergency mode after an error in the kdump environment occurs.
.PP
All files under /etc/kdump/emergency.d are collectively sorted
and executed in lexical order, before binary or script
specified kdump_emergency parameter is executed.
.PP
Note that scripts written for use with this directive must use the /bin/bash
interpreter. And since these scripts run in kdump enviroment, the reference to
the storage or network device in the scripts should adhere to the section
'Supported dump target types and requirements' in kexec-kdump-howto.txt.

.RE

.B kdump_post <binary | script>
.RS
This directive allows you to run a specified executable
just after the vmcore dump process terminates.  The exit
status of the current dump process is fed to the kdump_post
executable as its first argument($1).  Executable can modify
it to indicate the new exit status of succeeding dump process,
.PP
All files under /etc/kdump/post.d are collectively sorted
and executed in lexical order, before binary or script
specified kdump_post parameter is executed.
.PP
Note that scripts written for use with this directive must use the /bin/bash
interpreter. And since these scripts run in kdump enviroment, the reference to
the storage or network device in the scripts should adhere to the section
'Supported dump target types and requirements' in kexec-kdump-howto.txt.

.RE

.B kdump_pre <binary | script>
.RS
Works just like the "kdump_post" directive, but instead
of running after the dump process, runs immediately
before.  Exit status of this binary is interpreted
as follows:
.PP
0 - continue with dump process as usual
.PP
non 0 - run the final action (reboot/poweroff/halt)
.PP
All files under /etc/kdump/pre.d are collectively sorted and
executed in lexical order, after binary or script specified
kdump_pre parameter is executed.
Even if the binary or script in /etc/kdump/pre.d directory
returns non 0 exit status, the processing is continued.
.PP
Note that scripts written for use with this directive must use the /bin/bash
interpreter. And since these scripts run in kdump enviroment, the reference to
the storage or network device in the scripts should adhere to the section
'Supported dump target types and requirements' in kexec-kdump-howto.txt.

.RE

.B extra_bins <binaries | shell scripts>
.RS
This directive allows you to specify additional
binaries or shell scripts you'd like to include in
your kdump initrd. Generally only useful in
conjunction with a kdump_post binary or script that
relies on other binaries or scripts.
.RE

.B extra_modules <module(s)>
.RS
This directive allows you to specify extra kernel
modules that you want to be loaded in the kdump
initrd, typically used to set up access to
non-boot-path dump targets that might otherwise
not be accessible in the kdump environment. Multiple
modules can be listed, separated by spaces, and any
dependent modules will automatically be included.
.RE

.B failure_action <reboot | halt | poweroff | shell | dump_to_rootfs>
.RS
Action to perform in case dumping to the intended target fails. The default is "reboot".
reboot: Reboot the system (this is what most people will want, as it returns the system
to a normal state).  halt: Halt the system and lose the vmcore.  poweroff: The system
will be powered down.  shell: Drop to a shell session inside the initramfs, from which
you can manually perform additional recovery actions.  Exiting this shell reboots the
system by default or performs "final_action".
Note: kdump uses bash as the default shell.  dump_to_rootfs: If non-root dump
target is specified, the failure action can be set as dump_to_rootfs.  That means when
dumping to target fails, dump vmcore to rootfs from initramfs context and reboot
by default or perform "final_action".
.RE

.B default <reboot | halt | poweroff | shell | dump_to_rootfs>
.RS
Same as the "failure_action" directive above, but this directive is obsolete
and will be removed in the future.
.RE

.B final_action <reboot | halt | poweroff>
.RS
Action to perform in case dumping to the intended target succeeds.
Also performed when "shell" or "dump_to_rootfs" failure action finishes.
Each action is same as the "failure_action" directive above.
The default is "reboot".
.RE

.B force_rebuild <0 | 1>
.RS
By default, kdump initrd will only be rebuilt when necessary.
Specify 1 to force rebuilding kdump initrd every time when kdump service starts.
.RE

.B force_no_rebuild <0 | 1>
.RS
By default, kdump initrd will be rebuilt when necessary.
Specify 1 to bypass rebuilding of kdump initrd.

.PP
force_no_rebuild and force_rebuild options are mutually exclusive and
they should not be set to 1 simultaneously.
.RE

.B dracut_args <arg(s)>
.RS
Kdump uses dracut to generate initramfs for second kernel. This option
allows a user to pass arguments to dracut directly.
.RE


.B fence_kdump_args <arg(s)>
.RS
Command line arguments for fence_kdump_send (it can contain all valid
arguments except hosts to send notification to).
.RE


.B fence_kdump_nodes <node(s)>
.RS
List of cluster node(s) except localhost, separated by spaces, to send fence_kdump notification
to (this option is mandatory to enable fence_kdump).
.RE


.SH DEPRECATED OPTIONS

.B net <nfs mount>|<user@server>
.RS
net option is replaced by nfs and ssh options. Use nfs or ssh options
directly.
.RE

.B options <module> <option list>
.RS
Use KDUMP_COMMANDLINE_APPEND in /etc/sysconfig/kdump to add module options as
kernel command line parameters. For example, specify 'loop.max_loop=1' to limit
maximum loop devices to 1.
.RE

.B link_delay <seconds>
.RS
link_delay was used to wait for a network device to initialize before using it.
Now dracut network module takes care of this issue automatically.
.RE

.B disk_timeout <seconds>
.RS
Similar to link_delay, dracut ensures disks are ready before kdump uses them.
.RE

.B debug_mem_level <0-3>
.RS
Turn on verbose debug output of kdump scripts regarding free/used memory at
various points of execution.  This feature has been
moved to dracut now.
Use KDUMP_COMMANDLINE_APPEND in /etc/sysconfig/kdump and
append dracut cmdline param rd.memdebug=[0-3] to enable the debug output.

Higher level means more debugging output.
.PP
0 - no output
.PP
1 - partial /proc/meminfo
.PP
2 - /proc/meminfo
.PP
3 - /proc/meminfo + /proc/slabinfo
.RE

.B blacklist <list of kernel modules>
.RS
blacklist option was recently being used to prevent loading modules in
initramfs. General terminology for blacklist has been that module is
present in initramfs but it is not actually loaded in kernel. Hence
retaining blacklist option creates more confusing behavior. It has been
deprecated.
.PP
Instead, use rd.driver.blacklist option on second kernel to blacklist
a certain module. One can edit /etc/sysconfig/kdump.conf and edit
KDUMP_COMMANDLINE_APPEND to pass kernel command line options. Refer
to dracut.cmdline man page for more details on module blacklist option.
.RE

.RE

.SH EXAMPLES
Here are some examples for core_collector option:
.PP

changelog
* Thu Jul 18 2024 Fedora Release Engineering <releng@fedoraproject.org> - 1.0.44-2
- Rebuilt for https://fedoraproject.org/wiki/Fedora_41_Mass_Rebuild

* Mon Jul 15 2024 Packit <hello@packit.dev> - 1.0.44-1
- Update to version 1.0.44

* Tue Oct 24 2023 Coiby <coxu@redhat.com> - 1.0.42-10
- split from kexec-tools


dracut/99kdumpbase/kdump.sh
#!/bin/sh
#
# The main kdump routine in capture kernel, bash may not be the
# default shell. Any code added must be POSIX compliant.

# shellcheck source=/dev/null
. /lib/dracut-lib.sh
# shellcheck source=SCRIPTDIR/../../kdump-logger.sh
. /lib/kdump-logger.sh
# shellcheck source=SCRIPTDIR/../../kdump-lib-initramfs.sh
. /lib/kdump-lib-initramfs.sh

#initiate the kdump logger
if ! dlog_init; then
    echo "failed to initiate the kdump logger."
    exit 1
fi

KDUMP_PATH="/var/crash"
KDUMP_LOG_FILE="/run/initramfs/kexec-dmesg.log"
KDUMP_LOG_DEST=""
KDUMP_LOG_OP=""
KDUMP_TEST_ID=""
KDUMP_TEST_STATUS=""
CORE_COLLECTOR="makedumpfile -l --message-level 7 -d 31"
DMESG_COLLECTOR="/sbin/vmcore-dmesg"
FAILURE_ACTION="systemctl reboot -f"
DATEDIR=$(date +%Y-%m-%d-%T)
HOST_IP='127.0.0.1'
DUMP_INSTRUCTION=""
SSH_KEY_LOCATION=$DEFAULT_SSHKEY
DD_BLKSIZE=512
FINAL_ACTION="systemctl reboot -f"
KDUMP_PRE=""
KDUMP_POST=""
KDUMP_EMERGENCY=""
NEWROOT="/sysroot"
OPALCORE="/sys/firmware/opal/mpipl/core"
KDUMP_CONF_PARSED="/tmp/kdump.conf.$$"

# POSIX doesn't have pipefail, only apply when using bash
# shellcheck disable=SC3040
[ -n "$BASH" ] && set -o pipefail

DUMP_RETVAL=0

kdump_read_conf > $KDUMP_CONF_PARSED

get_kdump_confs() {
    while read -r config_opt config_val; do
        # remove inline comments after the end of a directive.
        case "$config_opt" in
            path)
                KDUMP_PATH="$config_val"
                ;;
            core_collector)
                [ -n "$config_val" ] && CORE_COLLECTOR="$config_val"
                ;;
            sshkey)
                if [ -f "$config_val" ]; then
                    SSH_KEY_LOCATION=$config_val
                fi
                ;;
            kdump_pre)
                KDUMP_PRE="$config_val"
                ;;
            kdump_post)
                KDUMP_POST="$config_val"
                ;;
            kdump_emergency)
                KDUMP_EMERGENCY="$config_val"
                ;;
            fence_kdump_args)
                FENCE_KDUMP_ARGS="$config_val"
                ;;
            fence_kdump_nodes)
                FENCE_KDUMP_NODES="$config_val"
                ;;
            failure_action | default)
                case $config_val in
                    shell)
                        FAILURE_ACTION="kdump_emergency_shell"
                        ;;
                    reboot)
                        FAILURE_ACTION="systemctl reboot -f && exit"
                        ;;
                    halt)
                        FAILURE_ACTION="halt && exit"
                        ;;
                    poweroff)
                        FAILURE_ACTION="systemctl poweroff -f && exit"
                        ;;
                    dump_to_rootfs)
                        FAILURE_ACTION="dump_to_rootfs"
                        ;;
                esac
                ;;
            final_action)
                case $config_val in
                    reboot)
                        FINAL_ACTION="systemctl reboot -f"
                        ;;
                    halt)
                        FINAL_ACTION="halt"
                        ;;
                    poweroff)
                        FINAL_ACTION="systemctl poweroff -f"
                        ;;
                esac
                ;;
        esac
    done < "$KDUMP_CONF_PARSED"

    case $CORE_COLLECTOR in
        *makedumpfile*)
            # Ensure no -F in makedumpfile by default.
            CORE_COLLECTOR=$(echo "$CORE_COLLECTOR" | sed -e "s/-F//g")
            if is_ssh_dump_target || is_raw_dump_target; then
                CORE_COLLECTOR="$CORE_COLLECTOR -F"
            fi
            THREADS=$(nproc)
            if [ "$THREADS" -gt 1 ]; then
                case "$CORE_COLLECTOR" in
                    *-F* | *-E*) ;;

                    *)
                        CORE_COLLECTOR="$CORE_COLLECTOR --num-threads=$THREADS"
                        ;;
                esac
            fi
            ;;
    esac

}

# store the kexec kernel log to a file.
save_log() {
    # LOG_OP is empty when log can't be saved, eg. raw target
    [ -n "$KDUMP_LOG_OP" ] || return

    dmesg -T > $KDUMP_LOG_FILE

    if command -v journalctl > /dev/null; then
        journalctl -ab >> $KDUMP_LOG_FILE
    fi
    chmod 600 $KDUMP_LOG_FILE

    dinfo "saving the $KDUMP_LOG_FILE to $KDUMP_LOG_DEST/"

    eval "$KDUMP_LOG_OP"
}

# $1: dump path, must be a mount point
dump_fs() {
    ddebug "dump_fs _mp=$1"

    if ! is_mounted "$1"; then
        dinfo "dump path '$1' is not mounted, trying to mount..."
        if ! mount --target "$1"; then
            derror "failed to dump to '$1', it's not a mount point!"
            return 1
        fi
    fi

    if [ -z "$KDUMP_TEST_ID" ]; then
        _dump_fs_path=$(echo "$1/$KDUMP_PATH/$HOST_IP-$DATEDIR/" | tr -s /)
    else
        _dump_fs_path=$(echo "$1/$KDUMP_PATH/" | tr -s /)
    fi

    dinfo "saving to $_dump_fs_path"

    # Only remount to read-write mode if the dump target is mounted read-only.
    _dump_mnt_op=$(get_mount_info OPTIONS target "$1" -f)
    case $_dump_mnt_op in
        ro*)
            dinfo "Remounting the dump target in rw mode."
            mount -o remount,rw "$1" || return 1
            ;;
    esac

    mkdir -p "$_dump_fs_path" || return 1

    save_vmcore_dmesg_fs ${DMESG_COLLECTOR} "$_dump_fs_path"
    save_opalcore_fs "$_dump_fs_path"

    dinfo "saving vmcore"
    KDUMP_LOG_DEST=$_dump_fs_path/
    KDUMP_LOG_OP="mv '$KDUMP_LOG_FILE' '$KDUMP_LOG_DEST/'"

    ddebug "Core Collector: $CORE_COLLECTOR"
    $CORE_COLLECTOR /proc/vmcore "$_dump_fs_path/vmcore-incomplete" > /dev/console 2>&1
    _dump_exitcode=$?
    if [ $_dump_exitcode -eq 0 ]; then
        sync -f "$_dump_fs_path/vmcore-incomplete"
        _sync_exitcode=$?
        if [ $_sync_exitcode -eq 0 ]; then
            mv "$_dump_fs_path/vmcore-incomplete" "$_dump_fs_path/vmcore"
            dinfo "saving vmcore complete"
        else
            derror "sync vmcore failed, exitcode:$_sync_exitcode"
            return 1
        fi
    else
        derror "saving vmcore failed, exitcode:$_dump_exitcode"
        return 1
    fi

    # improper kernel cmdline can cause the failure of echo, we can ignore this kind of failure
    return 0
}

# $1: dmesg collector
# $2: dump path
save_vmcore_dmesg_fs() {
    dinfo "saving vmcore-dmesg.txt to $2"
    if $1 /proc/vmcore > "$2/vmcore-dmesg-incomplete.txt"; then
        mv "$2/vmcore-dmesg-incomplete.txt" "$2/vmcore-dmesg.txt"
        chmod 600 "$2/vmcore-dmesg.txt"

        # Make sure file is on disk. There have been instances where later
        # saving vmcore failed and system rebooted without sync and there
        # was no vmcore-dmesg.txt available.
        sync
        dinfo "saving vmcore-dmesg.txt complete"
    else
        if [ -f "$2/vmcore-dmesg-incomplete.txt" ]; then
            chmod 600 "$2/vmcore-dmesg-incomplete.txt"
        fi
        derror "saving vmcore-dmesg.txt failed"
    fi
}

# $1: dump path
save_opalcore_fs() {
    if [ ! -f $OPALCORE ]; then
        # Check if we are on an old kernel that uses a different path
        if [ -f /sys/firmware/opal/core ]; then
            OPALCORE="/sys/firmware/opal/core"
        else
            return 0
        fi
    fi

    dinfo "saving opalcore:$OPALCORE to $1/opalcore"
    if ! cp $OPALCORE "$1/opalcore"; then
        derror "saving opalcore failed"
        return 1
    fi

    sync
    dinfo "saving opalcore complete"
    return 0
}

dump_to_rootfs() {

    if [ "$(systemctl status dracut-initqueue | sed -n "s/^\s*Active: \(\S*\)\s.*$/\1/p")" = "inactive" ]; then
        dinfo "Trying to bring up initqueue for rootfs mount"
        systemctl start dracut-initqueue
    fi

    dinfo "Clean up dead systemd services"
    systemctl cancel
    dinfo "Waiting for rootfs mount, will timeout after 90 seconds"
    systemctl start --no-block sysroot.mount

    _loop=0
    while [ $_loop -lt 90 ] && ! is_mounted /sysroot; do
        sleep 1
        _loop=$((_loop + 1))
    done

    if ! is_mounted /sysroot; then
        derror "Failed to mount rootfs"
        return
    fi

    ddebug "NEWROOT=$NEWROOT"
    dump_fs $NEWROOT
}

kdump_emergency_shell() {
    ddebug "Switching to kdump emergency shell..."

    # shellcheck source=/dev/null
    [ -f /etc/profile ] && . /etc/profile
    export PS1='kdump:${PWD}# '

    # shellcheck source=/dev/null
    . /lib/dracut-lib.sh
    if [ -f /dracut-state.sh ]; then
        # shellcheck source=/dev/null
        . /dracut-state.sh 2> /dev/null
    fi

    source_conf /etc/conf.d

    type plymouth > /dev/null 2>&1 && plymouth quit

    source_hook "emergency"
    while read -r _tty rest; do
        (
            echo
            echo
            echo 'Entering kdump emergency mode.'
            echo 'Type "journalctl" to view system logs.'
            echo 'Type "rdsosreport" to generate a sosreport, you can then'
            echo 'save it elsewhere and attach it to a bug report.'
            echo
            echo
        ) > "/dev/$_tty"
    done < /proc/consoles
    sh -i -l
    /bin/rm -f -- /.console_lock
}

do_failure_action() {
    dinfo "Executing failure action $FAILURE_ACTION"
    eval "$FAILURE_ACTION"
}

do_final_action() {
    dinfo "Executing final action $FINAL_ACTION"
    eval "$FINAL_ACTION"
}

do_dump() {
    eval "$DUMP_INSTRUCTION"
    _ret=$?

    if [ $_ret -ne 0 ]; then
        derror "saving vmcore failed"
    fi

    return $_ret
}

do_kdump_pre() {
    if [ -n "$KDUMP_PRE" ]; then
        "$KDUMP_PRE"
        _ret=$?
        if [ $_ret -ne 0 ]; then
            derror "$KDUMP_PRE exited with $_ret status"
            return $_ret
        fi
    fi

    # if any script fails, it just raises warning and continues
    if [ -d /etc/kdump/pre.d ]; then
        for file in /etc/kdump/pre.d/*; do
            "$file"
            _ret=$?
            if [ $_ret -ne 0 ]; then
                derror "$file exited with $_ret status"
            fi
        done
    fi
    return 0
}

do_kdump_post() {
    if [ -d /etc/kdump/post.d ]; then
        for file in /etc/kdump/post.d/*; do
            "$file" "$1"
            _ret=$?
            if [ $_ret -ne 0 ]; then
                derror "$file exited with $_ret status"
            fi
        done
    fi

    if [ -n "$KDUMP_POST" ]; then
        "$KDUMP_POST" "$1"
        _ret=$?
        if [ $_ret -ne 0 ]; then
            derror "$KDUMP_POST exited with $_ret status"
        fi
    fi
}

do_kdump_emergency() {
    if [ -d /etc/kdump/emergency.d ]; then
        for file in /etc/kdump/emergency.d/*; do
            "$file"
            _ret=$?
            if [ $_ret -ne 0 ]; then
                derror "$file exited with $_ret status"
            fi
        done
    fi

    if [ -n "$KDUMP_EMERGENCY" ]; then
        "$KDUMP_EMERGENCY"
        _ret=$?
        if [ $_ret -ne 0 ]; then
            derror "$KDUMP_EMERGENCY exited with $_ret status"
        fi
    fi
}
# $1: block target, eg. /dev/sda
dump_raw() {
    [ -b "$1" ] || return 1

    dinfo "saving to raw disk $1"

    if ! echo "$CORE_COLLECTOR" | grep -q makedumpfile; then
        _src_size=$(stat --format %s /proc/vmcore)
        _src_size_mb=$((_src_size / 1048576))
        /kdumpscripts/monitor_dd_progress.sh $_src_size_mb &
    fi

    dinfo "saving vmcore"
    ddebug "Core Collector: $CORE_COLLECTOR"
    $CORE_COLLECTOR /proc/vmcore | dd of="$1" bs=$DD_BLKSIZE >> /tmp/dd_progress_file 2>&1 || return 1
    sync

    dinfo "saving vmcore complete"
    return 0
}

# $1: ssh key file
# $2: ssh address in <user>@<host> format
dump_ssh() {
    _ret=0
    _ssh_opts="-i $1 -o BatchMode=yes -o StrictHostKeyChecking=yes"
    if [ -z "$KDUMP_TEST_ID" ]; then
        _ssh_dir="$KDUMP_PATH/$HOST_IP-$DATEDIR"
    else
        _ssh_dir="$KDUMP_PATH"
    fi

    if is_ipv6_address "$2"; then
        _scp_address=${2%@*}@"[${2#*@}]"
    else
        _scp_address=$2
    fi

    dinfo "saving to $2:$_ssh_dir"

    cat /var/lib/random-seed > /dev/urandom
    # shellcheck disable=SC2086 # ssh_opts needs to be split
    ssh -q $_ssh_opts "$2" mkdir -p "$_ssh_dir" || return 1

    save_vmcore_dmesg_ssh "$DMESG_COLLECTOR" "$_ssh_dir" "$_ssh_opts" "$2"

    dinfo "saving vmcore"

    KDUMP_LOG_DEST=$2:$_ssh_dir/
    KDUMP_LOG_OP="scp -q $_ssh_opts '$KDUMP_LOG_FILE' '$_scp_address:$_ssh_dir/'"

    save_opalcore_ssh "$_ssh_dir" "$_ssh_opts" "$2" "$_scp_address"

    if [ "${CORE_COLLECTOR%%[[:blank:]]*}" = "scp" ]; then
        # shellcheck disable=SC2086 # ssh_opts needs to be split
        scp -q $_ssh_opts /proc/vmcore "$_scp_address:$_ssh_dir/vmcore-incomplete"
        _ret=$?
        _vmcore="vmcore"
    else
        # shellcheck disable=SC2029,SC2086
        #  - _ssh_opts needs to be split
        #  - _ssh_dir needs to be expanded
        ddebug "Core Collector: $CORE_COLLECTOR"
        $CORE_COLLECTOR /proc/vmcore | ssh $_ssh_opts "$2" "umask 0077 && dd bs=512 of='$_ssh_dir/vmcore-incomplete'"
        _ret=$?
        _vmcore="vmcore.flat"
    fi

    if [ $_ret -eq 0 ]; then
        # shellcheck disable=SC2029,SC2086
        #  - _ssh_opts needs to be split
        #  - _ssh_dir needs to be expanded
        ssh $_ssh_opts "$2" "mv '$_ssh_dir/vmcore-incomplete' '$_ssh_dir/$_vmcore'"
        _ret=$?
        if [ $_ret -ne 0 ]; then
            derror "moving vmcore failed, exitcode:$_ret"
        else
            dinfo "saving vmcore complete"
        fi
    else
        derror "saving vmcore failed, exitcode:$_ret"
    fi

    return $_ret
}

# $1: dump path
# $2: ssh opts
# $3: ssh address in <user>@<host> format
# $4: scp address, similar with ssh address but IPv6 addresses are quoted
save_opalcore_ssh() {
    if [ ! -f $OPALCORE ]; then
        # Check if we are on an old kernel that uses a different path
        if [ -f /sys/firmware/opal/core ]; then
            OPALCORE="/sys/firmware/opal/core"
        else
            return 0
        fi
    fi

    dinfo "saving opalcore:$OPALCORE to $3:$1"

    # shellcheck disable=SC2086 # $2 (_ssh_opts) needs to be split
    if ! scp $2 "$OPALCORE" "$4:$1/opalcore-incomplete"; then
        derror "saving opalcore failed"
        return 1
    fi

    # shellcheck disable=SC2029,SC2086
    #  - $1 (dump path) needs to be expanded
    #  - $2 (_ssh_opts) needs to be split
    ssh $2 "$3" mv "$1/opalcore-incomplete" "$1/opalcore"
    dinfo "saving opalcore complete"
    return 0
}

# $1: dmesg collector
# $2: dump path
# $3: ssh opts
# $4: ssh address in <user>@<host> format
save_vmcore_dmesg_ssh() {
    dinfo "saving vmcore-dmesg.txt to $4:$2"
    # shellcheck disable=SC2029,SC2086
    #  - $2 (_ssh_dir) needs to be expanded
    #  - $3 (_ssh_opts) needs to be split
    if "$1" /proc/vmcore | ssh $3 "$4" "umask 0077 && dd of='$2/vmcore-dmesg-incomplete.txt'"; then
        ssh -q $3 "$4" mv "$2/vmcore-dmesg-incomplete.txt" "$2/vmcore-dmesg.txt"
        dinfo "saving vmcore-dmesg.txt complete"
    else
        derror "saving vmcore-dmesg.txt failed"
    fi
}

wait_online_network() {
    # In some cases, network may still not be ready because nm-online is called
    # with "-s" which means to wait for NetworkManager startup to complete, rather
    # than waiting for network connectivity specifically. Wait 10mins more for the
    # network to be truely ready in these cases.
    _loop=0
    while [ $_loop -lt 600 ]; do
        sleep 1
        _loop=$((_loop + 1))
        if _route=$(kdump_get_ip_route "$1" 2> /dev/null); then
            printf "%s" "$_route"
            return
        else
            dwarn "Waiting for network to be ready (${_loop}s / 10min)"
        fi
    done

    derror "Oops. The network still isn't ready after waiting 10mins."
    exit 1
}

get_host_ip() {

    if ! is_nfs_dump_target && ! is_ssh_dump_target; then
        return 0
    fi

    _kdump_remote_ip=$(getarg kdump_remote_ip=)

    if [ -z "$_kdump_remote_ip" ]; then
        derror "failed to get remote IP address!"
        return 1
    fi

    if ! _route=$(wait_online_network "$_kdump_remote_ip"); then
        return 1
    fi

    _netdev=$(kdump_get_ip_route_field "$_route" "dev")

    if ! _kdumpip=$(ip addr show dev "$_netdev" | grep '[ ]*inet'); then
        derror "Failed to get IP of $_netdev"
        return 1
    fi

    _kdumpip=$(echo "$_kdumpip" | head -n 1 | awk '{print $2}')
    _kdumpip="${_kdumpip%%/*}"
    HOST_IP=$_kdumpip
}

read_kdump_confs() {
    if [ ! -f "$KDUMP_CONFIG_FILE" ]; then
        derror "$KDUMP_CONFIG_FILE not found"
        return
    fi

    get_kdump_confs

    # rescan for add code for dump target
    while read -r config_opt config_val; do
        # remove inline comments after the end of a directive.
        case "$config_opt" in
            dracut_args)
                config_val=$(get_dracut_args_target "$config_val")
                if [ -n "$config_val" ]; then
                    config_val=$(get_mntpoint_from_target "$config_val")
                    DUMP_INSTRUCTION="dump_fs $config_val"
                fi
                ;;
            ext[234] | xfs | btrfs | minix | nfs | virtiofs)
                config_val=$(get_mntpoint_from_target "$config_val")
                DUMP_INSTRUCTION="dump_fs $config_val"
                ;;
            raw)
                DUMP_INSTRUCTION="dump_raw $config_val"
                ;;
            ssh)
                DUMP_INSTRUCTION="dump_ssh $SSH_KEY_LOCATION $config_val"
                ;;
        esac
    done < "$KDUMP_CONF_PARSED"
}

fence_kdump_notify() {
    if [ -n "$FENCE_KDUMP_NODES" ]; then
        # shellcheck disable=SC2086
        $FENCE_KDUMP_SEND $FENCE_KDUMP_ARGS $FENCE_KDUMP_NODES &
    fi
}

kdump_test_set_status() {
    _status="$1"

    [ -n "$KDUMP_TEST_STATUS" ] || return

    case "$_status" in
        success | fail) ;;
        *)
            derror "Unknown test status $_status"
            return 1
            ;;
    esac

    if is_ssh_dump_target; then
        _ssh_opts="-i $SSH_KEY_LOCATION -o BatchMode=yes -o StrictHostKeyChecking=yes"
        _ssh_host=$(echo "$DUMP_INSTRUCTION" | awk '{print $3}')

        # shellcheck disable=SC2086 # need to word-split $_ssh_opts
        ssh -q $_ssh_opts "$_ssh_host" "mkdir -p ${KDUMP_TEST_STATUS%/*}" \
            || return 1
        # shellcheck disable=SC2086 # need to word-split $_ssh_opts
        ssh -q $_ssh_opts "$_ssh_host" "echo $_status kdump_test_id=$KDUMP_TEST_ID > $KDUMP_TEST_STATUS" \
            || return 1
    else
        _target=$(echo "$DUMP_INSTRUCTION" | awk '{print $2}')

        mkdir -p "$_target/$KDUMP_PATH" || return 1
        echo "$_status kdump_test_id=$KDUMP_TEST_ID" > "$_target/$KDUMP_TEST_STATUS"
        sync -f "$_target/$KDUMP_TEST_STATUS"
    fi
}

kdump_test_init() {
    is_raw_dump_target && return

    KDUMP_TEST_ID=$(getarg kdump_test_id=)
    [ -z "$KDUMP_TEST_ID" ] && return

    KDUMP_PATH="$KDUMP_PATH/kdump-test-$KDUMP_TEST_ID"
    KDUMP_TEST_STATUS="$KDUMP_PATH/vmcore-creation.status"

    kdump_test_set_status 'fail'
}

if [ "$1" = "--error-handler" ]; then
    get_kdump_confs
    do_kdump_emergency
    do_failure_action
    save_log
    do_final_action

    exit $?
fi

# continue here only if we have to save dump.
if [ -f /etc/fadump.initramfs ] && [ ! -f /proc/device-tree/rtas/ibm,kernel-dump ] && [ ! -f /proc/device-tree/ibm,opal/dump/mpipl-boot ]; then
    exit 0
fi

read_kdump_confs
fence_kdump_notify

if ! get_host_ip; then
    derror "get_host_ip exited with non-zero status!"
    exit 1
fi

if [ -z "$DUMP_INSTRUCTION" ]; then
    DUMP_INSTRUCTION="dump_fs $NEWROOT"
fi

kdump_test_init
if ! do_kdump_pre; then
    derror "kdump_pre script exited with non-zero status!"
    do_final_action
    # During systemd service to reboot the machine, stop this shell script running
    exit 1
fi
make_trace_mem "kdump saving vmcore" '1:shortmem' '2+:mem' '3+:slab'
do_dump
DUMP_RETVAL=$?

if ! do_kdump_post $DUMP_RETVAL; then
    derror "kdump_post script exited with non-zero status!"
fi

save_log

if [ $DUMP_RETVAL -ne 0 ]; then
    exit 1
fi

kdump_test_set_status "success"
do_final_action
